### PR TITLE
Implement CHARLEN and CHARSUB

### LINKED ISSUE #786: [Feature request] Functions for working with charmapped values from strings
- `CHARLEN("string")` returns the length of `"string"` in charmap entries
- `CHARSUB("string", n)` returns the *n*th charmap substring of `"string"`

This would make `db/dw/dl "{S}"` equivalent to:

```
for n, 1, CHARLEN("{S}") + 1
db/dw/dl CHARSUB("{S}", n)
endr
```

For example:

```
    charmap "A", $10
    charmap "B", $20
    charmap "C", $30
    charmap "Bold", $88

    assert CHARLEN("XBoldABC") == 5
    assert !STRCMP(CHARSUB("XBoldABC", 2), "Bold")
    assert CHARSUB("XBoldABC", 2) == "Bold" ; numeric conversion makes them both $88
    assert CHARSUB("XBoldABC", 1) == $58 ; ASCII for "X"
    println CHARSUB("XBoldABC", 2) ; prints "Bold"
```

If charmaps gain multi-byte support, a third function would also be useful, to get the length of a charmapped string in output bytes.

### PR DESCRIPTION
**Pull request by Rangi42**
Fixes #786

### CODE REVIEW

**Comment by Rangi42** [comment]
@ISSOtm Refactoring `*output` to `**output` is in a separate commit in case you prefer the former.

File: src/asm/charmap.c [hunk 1]
```diff
@@ -216,31 +225,35 @@ size_t charmap_Convert(char const *input, uint8_t *output)
...
+				return codepointLen;
+
+			} else { /* End of input */
+				return 0;
+			}
 		}
 	}
 
-	return outputLen;
+	unreachable_();
```
Comment (ISSOtm) [inline src/asm/charmap.c:258]: Is this really necessary? The for loop is infinite and cannot be exited from, can it?

File: src/asm/rgbasm.5 [hunk 2]
```diff
@@ -399,6 +399,8 @@ Most of them return a string, however some of these functions actually return an
 .It Fn STRLWR str Ta Returns Ar str No with all letters in lowercase.
 .It Fn STRRPL str old new Ta Returns Ar str No with each non-overlapping occurrence of the substring Ar old No replaced with Ar new .
 .It Fn STRFMT fmt args... Ta Returns the string Ar fmt No with each
+.It Fn CHARLEN str Ta Returns the number of charmap entries in Ar str No with the current charmap.
+.It Fn CHARSUB str pos Ta Returns the substring for the charmap entry at Ar pos No in Ar str No Po first character is position 1 Pc with the current charmap.
```
Comment (ISSOtm) [inline src/asm/rgbasm.5:403]: No need for `Po` and `Pc`, regular parens are fine.

**Review by ISSOtm — CHANGES_REQUESTED** [review]

File: src/asm/charmap.c (hunk 1 above)
Comment (Rangi42) [reply src/asm/charmap.c:258]: I must have added it because at least one CI build complained without it.

File: src/asm/rgbasm.5 (hunk 2 above)
Comment (Rangi42) [reply src/asm/rgbasm.5:403]: Oh, I was copying `STRSUB`'s line.

File: src/asm/rgbasm.5 [hunk 3]
```diff
@@ -394,11 +394,13 @@ Most of them return a string, however some of these functions actually return an
 .It Fn STRCMP str1 str2 Ta Returns -1 if Ar str1 No is alphabetically lower than Ar str2 No , zero if they match, 1 if Ar str1 No is greater than Ar str2 .
 .It Fn STRIN str1 str2 Ta Returns the first position of Ar str2 No in Ar str1 No or zero if it's not present Pq first character is position 1 .
 .It Fn STRRIN str1 str2 Ta Returns the last position of Ar str2 No in Ar str1 No or zero if it's not present Pq first character is position 1 .
-.It Fn STRSUB str pos len Ta Returns a substring from Ar str No starting at Ar pos Po first character is position 1 Pc and Ar len No characters long.
+.It Fn STRSUB str pos len Ta Returns a substring from Ar str No starting at Ar pos (first character is position 1) and Ar len No characters long.
```
Comment (ISSOtm) [inline src/asm/rgbasm.5:397]: `Po` was actually being (incorrectly, I think) used to cancel the `Ar`; we need a `No` before the opening parens.

**Review by ISSOtm — CHANGES_REQUESTED** [review]

**Review by ISSOtm — APPROVED** [review]

**Pull request merged**